include/CL/opencl.hpp

```diff
@@ -10253,7 +10370,118 @@ namespace compatibility {
+    {
+        typedef CL_API_ENTRY cl_int(CL_API_CALL * PFN_clGetSemaphoreInfoKHR)(
+            cl_semaphore_khr sema_object, cl_semaphore_info_khr param_name,
+            size_type param_value_size, void* param_value,
+            size_type* param_value_size_ret);
+
+        return detail::errHandler(
+            detail::getInfo(&pfn_clGetSemaphoreInfoKHR, object_, name, param),
+            __GET_SEMAPHORE_KHR_INFO_ERR);
+    }
+    template <cl_semaphore_info_khr name> typename
+    detail::param_traits<detail::cl_semaphore_info_khr, name>::param_type
+    getInfo(cl_int* err = NULL) const
+    {
+        typename detail::param_traits<
+            detail::cl_semaphore_info_khr, name>::param_type param;
+        cl_int result = getInfo(name, &param);
+        if (err != NULL) {
+            *err = result;        
+        }
+        return param;      
+    }
+
+    cl_int retain()
+    { 
+        return pfn_clRetainSemaphoreKHR(object_);
+    }
+
+    cl_int release()
+    { 
+        return pfn_clReleaseSemaphoreKHR(object_);
+    }
+
+private:
+    static std::once_flag ext_init_;
+    static cl_int ext_init_error_;
+
+    static void initExtensions(const Context& context)
+    {
+#if CL_HPP_TARGET_OPENCL_VERSION >= 120
+        Device device = context.getInfo<CL_CONTEXT_DEVICES>().at(0);
+        cl_platform_id platform = device.getInfo<CL_DEVICE_PLATFORM>();
+        CL_HPP_INIT_CL_EXT_FCN_PTR_PLATFORM_(platform, clCreateSemaphoreWithPropertiesKHR);
+        CL_HPP_INIT_CL_EXT_FCN_PTR_PLATFORM_(platform, clReleaseSemaphoreKHR);
+        CL_HPP_INIT_CL_EXT_FCN_PTR_PLATFORM_(platform, clRetainSemaphoreKHR);
+        CL_HPP_INIT_CL_EXT_FCN_PTR_PLATFORM_(platform, clEnqueueWaitSemaphoresKHR);
+        CL_HPP_INIT_CL_EXT_FCN_PTR_PLATFORM_(platform, clEnqueueSignalSemaphoresKHR);
+        CL_HPP_INIT_CL_EXT_FCN_PTR_PLATFORM_(platform, clGetSemaphoreInfoKHR);
+#else
+        CL_HPP_INIT_CL_EXT_FCN_PTR_(clCreateSemaphoreWithPropertiesKHR);
+        CL_HPP_INIT_CL_EXT_FCN_PTR_(clReleaseSemaphoreKHR);
+        CL_HPP_INIT_CL_EXT_FCN_PTR_(clRetainSemaphoreKHR);
+        CL_HPP_INIT_CL_EXT_FCN_PTR_(clEnqueueWaitSemaphoresKHR);
+        CL_HPP_INIT_CL_EXT_FCN_PTR_(clEnqueueSignalSemaphoresKHR);
+        CL_HPP_INIT_CL_EXT_FCN_PTR_(clGetSemaphoreInfoKHR);
+
+#endif
+        if ((pfn_clCreateSemaphoreWithPropertiesKHR == nullptr) ||
+            (pfn_clReleaseSemaphoreKHR              == nullptr) ||
+            (pfn_clRetainSemaphoreKHR               == nullptr) ||
+            (pfn_clEnqueueWaitSemaphoresKHR         == nullptr) ||
+            (pfn_clEnqueueSignalSemaphoresKHR       == nullptr) ||
+            (pfn_clGetSemaphoreInfoKHR              == nullptr))
+        {
+            ext_init_error_ = CL_INVALID_VALUE;
+        }
+    }
+
+};
+
+CL_HPP_DEFINE_STATIC_MEMBER_ std::once_flag Semaphore::ext_init_;
+CL_HPP_DEFINE_STATIC_MEMBER_ cl_int Semaphore::ext_init_error_ = CL_SUCCESS;
```

Comment:
I think I understand why this static error variable exists (it ensures a semaphore cannot be created unless all function pointers are properly queried) and I'm OK having it, but an alternative we could consider is to allow `nullptr` function pointers and only return an error if/when trying to use a `nullptr` function pointer.  What do you think?

Reply:
The only purpose for these variables is to satisfy the call_once for `InitExtensions`. `initExtensions` is called in the constructor and one of the pointers (`pfn_clCreateSemaphoreWithPropertiesKHR`) is called after that, so we would need to throw the exception there anyway or shall we skip the function call? 
I'm wondering if there could be a scenario where someone would want to call it for multiple devices (e.g. in a loop trying to find one that supports semaphores).

Reply:
Good questions, I think this discussion really amounts to: how do we want to behave if one or more semaphore function pointers cannot be queried?

Option 1: `initExtensions` succeeds but sets a static variable to an error condition.  Any attempts to create a semaphore fail.  This is the current behavior.

Option 2: `initExtensions` succeeds, but any attempts to use a `nullptr` function pointer will fail.  This might allow some semaphore functionality, depending on which function pointers are `nullptr`.  This was my alternative suggested above.

Option 3: `initExtensions` fails (throws an exception) and the attempt to create a semaphore fails.  Trying to create a semaphore with a different context with a different device may succeed.  This is a little bit like Option 1, but with the retry behavior.

Note that if we want to go with Option 3 we will need to throw an exception from `initExtension`, otherwise `std::call_once` will not try again:  https://en.cppreference.com/w/cpp/thread/call_once

Reply:
Thanks for the note. I wasn't aware of that :)
Regarding option 2, is it possible that not all pointers are correctly initialized for some platforms? I don't know all the usage scenarios, but options 2 and 3 look fine to me. Which one do you prefer?

Reply:
Let's go with Option 2 for now.  I think this is most consistent with the extension function pointer usage elsewhere in this file.  I'll open a separate issue to revisit extension handling in general, but we can solve that with separate PRs.

So, specific tasks:
* Remove all of the `nullptr` checks from `initExtensions`.
* Remove `ext_init_error_`.
* Add a check for a `nullptr` `pfn_clCreateSemaphoreWithPropertiesKHR` in the `Semaphore` constructor, since we will no longer be checking against `ext_init_error_`.

Sound good?

For "bonus points" we could throw an exception and retry if _none_ of the extension function pointers are non-`nullptr`, but this is probably overkill.

Reply:
Thanks!  Latest update looks great 👍